src/dynamics/poincare_recurrence.lean

```diff
@@ -0,0 +1,151 @@
+import measure_theory.measure_space
+
+lemma equiv.perm.coe_pow {α : Type*} (e : equiv.perm α) : ∀ n : ℕ, ⇑(e ^ n) = (e^[n])
+| 0 := rfl
+| (n+1) := by { ext x, rw [nat.iterate_succ, pow_succ', equiv.perm.mul_apply, equiv.perm.coe_pow] }
+
+lemma nat.iterate_mul {α : Sort*} (f : α → α) (m : ℕ) :
+  ∀ n, f^[m * n] = (f^[m]^[n])
+| 0 := by { ext x, simp }
+| (n + 1) := by { ext x, simp [mul_add, nat.iterate_add, nat.iterate_mul n] }
+
+lemma exists_imp_distrib' {α : Sort*} {p q : α → Prop} :
+  (∃ x, p x → q x) ↔ ((∀ x, p x) → ∃ x, q x) :=
+by classical; simp only [imp_iff_not_or, exists_or_distrib, not_forall]
+
+lemma forall_imp_distrib_left {α : Sort*} {p : Prop} {q : α → Prop} [decidable p] :
+  (∀ x, p → q x) ↔ (p → ∀ x, q x) :=
+by simp only [imp_iff_not_or, forall_or_distrib_left]
+
+noncomputable theory
+
+open classical set filter measure_theory finset
+open_locale classical topological_space
+
+variables {ι : Type*} {α : Type*} {β : Type*} [measure_space α]
+
+lemma sum_volume_le_volume_univ {s : finset ι} {t : ι → set α} (h : ∀ i ∈ s, is_measurable (t i))
+  (H : pairwise_on ↑s (disjoint on t)) : s.sum (λ i, volume (t i)) ≤ volume (univ : set α) :=
+volume_bUnion_finset H h ▸ volume_mono (subset_univ _)
+
+lemma tsum_volume_le_volume_univ {s : ι → set α} (hs : ∀ i, is_measurable (s i))
+  (H : pairwise (disjoint on s)) :
+  (∑ i, volume (s i)) ≤ volume (univ : set α) :=
+begin
+  rw [ennreal.tsum_eq_supr_sum],
+  exact supr_le (λ s, sum_volume_le_volume_univ (λ i hi, hs i) (λ i hi j hj hij, H i j hij))
+end
+
+/-- Pigeonhole principle for measure spaces: if `∑ i, μ (s i) > μ univ`, then
+one of the intersections `s i ∩ s j` is not empty. -/
+lemma exists_nonempty_inter_of_volume_univ_lt_tsum_volume {s : ι → set α}
+  (hs : ∀ i, is_measurable (s i)) (H : volume (univ : set α) < ∑ i, volume (s i)) :
+  ∃ i j (h : i ≠ j), (s i ∩ s j).nonempty :=
+begin
+  contrapose! H,
+  apply tsum_volume_le_volume_univ hs,
+  exact λ i j hij x hx, H i j hij ⟨x, hx⟩
+end
+
+/-- Pigeonhole principle for measure spaces: if `s` is a `finset` and
+`s.sum (λ i, μ (t i)) > μ univ`, then one of the intersections `t i ∩ t j` is not empty. -/
+lemma exists_nonempty_inter_of_volume_univ_lt_sum_volume {s : finset ι} {t : ι → set α}
+  (h : ∀ i ∈ s, is_measurable (t i)) (H : volume (univ : set α) < s.sum (λ i, volume (t i))) :
+  ∃ (i ∈ s) (j ∈ s) (h : i ≠ j), (t i ∩ t j).nonempty :=
+begin
+  contrapose! H,
+  apply sum_volume_le_volume_univ h,
+  exact λ i hi j hj hij x hx, H i hi j hj hij ⟨x, hx⟩
+end
+
+lemma measurable.iterate {f : α → α} (hf : measurable f) : ∀ n, measurable (f^[n])
+| 0 := measurable_id
+| (n+1) := (measurable.iterate n).comp hf
+
+/-- A measurable self-map `f` is called `volume_preserving` if for any measurable `s`
+we have `μ (f ⁻¹' s) = μ s`, or equivalently `μ.map f = μ`. -/
+def volume_preserving (f : α → α) : Prop :=
```

Comment:
Do you want to have a predicate like that, or to bundle measure preserving maps between measure spaces (just like we bundle group morphisms, and so on)?

Reply:
I don't know yet. Do you have any preferences?

Reply:
I think I would prefer bundling, because it means you don't need to invoke the volume preserving assumption every time you apply a theorem, and also that you get it for free for iterates of the map. Another option would be to have a typeclass like `[fact (is_volume_preserving f)]`, but we have seen with groups and everything that this is not a good design.

Reply:
In fact maybe we could do it with type classes, to have a hierarchy with volume preserving, ergodic, weakly mixing, mixing, aware of facts that the powers of a mixing transformation are ergodic and things like that. With the rule that we *don't* put the fact that a composition of measure preserving maps is measure preserving in the type classes.

What do you think?

Reply:
I think that I prefer typeclasses because most probably `[quasi_volume_preserving f]` will not be the most important property of a map `f`. I wonder what should we do with `measurable`.

Reply:
I agree with you for the type classes. For measurable, I wouldn't put it as type class, but there should be a lemma (with a typeclass assumption) saying that a quasi-volume preserving map is measurable. 

Ideally, we would have a `measurable_tac` tactic that could check measurability from the context and the usual lemmas, and that would be a default tactic in all lemmas assuming measurability (I mean, the measurability assumptions would be stated like `(h : measurable u . measurable_tac)`, and then it would be discharged automatically in most cases).

Reply:
I decided to go with explicit unbundled predicates. Let's see if it works well.